diff --git a/src/libhydra/plugins/kernel_pfkey/kernel_pfkey_ipsec.c b/src/libhydra/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
index 4704d41..4f4a1bc 100644
--- a/src/libhydra/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
+++ b/src/libhydra/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
@@ -33,6 +33,7 @@
 #endif
 
 #ifdef HAVE_NETIPSEC_IPSEC_H
+#include <stdint.h>
 #include <netipsec/ipsec.h>
 #elif defined(HAVE_NETINET6_IPSEC_H)
 #include <netinet6/ipsec.h>
