diff --git a/debian/patches/03_systemd-service.patch b/debian/patches/03_systemd-service.patch
diff --git a/init/systemd/strongswan.service.in b/init/systemd/strongswan.service.in
index 474284a19..8060d1ea2 100644
--- a/init/systemd/strongswan.service.in
+++ b/init/systemd/strongswan.service.in
@@ -1,9 +1,10 @@
 [Unit]
 Description=strongSwan IPsec IKEv1/IKEv2 daemon using ipsec.conf
-After=syslog.target network-online.target
+After=network-online.target
 
 [Service]
 ExecStart=@SBINDIR@/@IPSEC_SCRIPT@ start --nofork
+ExecReload=@SBINDIR@/@IPSEC_SCRIPT@ reload
 StandardOutput=syslog
 Restart=on-abnormal
 
